iceaddr/addresses.py

```diff
@@ -51,27 +51,33 @@ def iceaddr_lookup(
     # Look up postcodes for placename if no postcode is provided
     if placename and not postcode:
         pc = postcodes_for_placename(placename.strip())
+    query = "SELECT * FROM stadfong WHERE"
+    name_fields = ["heiti_nf=?", "heiti_tgf=?"]
+    if not number:
+        # Add lookup for churches and places of interest like Harpa
+        name_fields.append("serheiti=?")
+    query += "({})".format(" OR ".join(name_fields))
+    args = [street_name] * len(name_fields)
 
-    q = "SELECT * FROM stadfong WHERE (heiti_nf=? OR heiti_tgf=?)"
-    sqlargs = [street_name, street_name]
     if number:
-        q += " AND husnr=? "
-        sqlargs.append(number)
+        query += " AND (husnr=? OR substr(vidsk, 0, instr(vidsk, '-')) = ?)"
+        args.extend((number, str(number)))
         if letter:
-            q += " AND bokst LIKE ? COLLATE NOCASE"
-            sqlargs.append(letter)
+            query += " AND bokst LIKE ? COLLATE NOCASE"
+            args.append(letter)
+
     if pc:
         qp = " OR ".join([" postnr=?" for p in pc])
-        sqlargs.extend(pc)
-        q += " AND (%s) " % qp
+        args.extend(pc)
+        query += " AND (%s) " % qp
 
     # Ordering by postcode may in fact be a reasonable proxy
     # for delivering by order of match likelihood since the
     # lowest postcodes are generally more densely populated
-    q += " ORDER BY postnr ASC, husnr ASC, bokst ASC LIMIT ?"
-    sqlargs.append(limit)
+    query += " ORDER BY vidsk != '', postnr ASC, husnr ASC, bokst ASC LIMIT ?"
```

Comment:
If there are matches without `vidsk` we want those first so that `husnr` matches take precedence.